linter/linter.go

```diff
@@ -34,7 +34,17 @@ func Check(tbs *x509.Certificate, subjectPubKey crypto.PublicKey, realIssuer *x5
 	if err != nil {
 		return err
 	}
-	_, err = linter.Check(tbs, subjectPubKey)
+
+	// If the provided pubkey is identical to the public half of the provided
+	// signing key, then we're being asked to lint a self-signed cert. Replace the
+	// pubkey with the fake lint issuer's pubkey, so that the resulting lint cert
+	// is also self-signed.
+	lintSubjectPubKey := subjectPubKey
+	if realSigner.Public() == subjectPubKey {
```

Comment:
This is not doing what it looks like it's doing. `Public()` returns `any`, which is an interface value.

https://go.dev/ref/spec#Comparison_operators

> Interface types that are not type parameters are comparable. Two interface values are equal if they have [identical](https://go.dev/ref/spec#Type_identity) dynamic types and equal dynamic values or if both have value nil.

> Pointer types are comparable. Two pointer values are equal if they point to the same variable or if both have value nil. Pointers to distinct [zero-size](https://go.dev/ref/spec#Size_and_alignment_guarantees) variables may or may not be equal.

We happen to know that the likely values for `Public()` are `*ecdsa.PublicKey` and `*rsa.PublicKey`. Since those are pointers, this comparison will do a pointer equality check, rather than a check of the internal bytes.

To compare the _contents_, which is the less surprising thing, we could do one of two things:

 - marshal both sides and compare the bytes. That's what we do in `core.PublicKeysEqual`.
 - do a type switch and call the [respective](https://pkg.go.dev/crypto/ecdsa#PublicKey.Equal) equality [functions](https://pkg.go.dev/crypto/rsa#PublicKey.Equal) (I like this a little better and maybe we should update core.PublicKeysEqual to do it).

Reply:
Yeah, I thought about this, and figured the pointer comparison was good enough because it will never think a cert is self-signed when it actually isn't. It can fail in the other direction (thinking a cert isn't self-signed when it should be) but that will mostly cause _more_ Sub-CA lints to come into effect. And in all of our use-cases, the pointer comparison works.

But yes for sure, using a type switch and using `<specific pubkey type>.Equal()` is the more correct thing to do. I just didn't like the way every clause of the type switch was going to have the same body. Something about that felt like there should be a better way to do it.

Anyway, done, let me know if you see a more elegant way to do this.

Reply:
Also, I filed https://github.com/letsencrypt/boulder/pull/6996 for doing the same to core.PublicKeysEqual

Reply:
Yeah, I suspected that pointer comparison would work in all our use cases. But this just felt like a really surprising bug waiting to bite someone. Like if some code path changed so a pubkey was round-tripped through a parsing step, this would break in a mysterious way.